### PR TITLE
Added Documentation for Binary Search Tree

### PR DESCRIPTION
**Pull request by Saatvik21**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

`Yes`

### Description

I've added the docs for BST.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Fixes : #68

### CODE REVIEW

File: docs/data-structures/binary-search-tree.mdx [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+1: Insert
+2: Search
+3: Delete
+4: Print (Inorder)
+5: Exit
+Enter option: 2
+Enter element to search for: 10
+```
+
+### Sample Output
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:464]: Put a line break

File: docs/data-structures/binary-search-tree.mdx [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+- Insert : O(n)
+- Search : O(n)
+- Delete : O(n)
+- Print (Inorder): O(n)
+
+## Space Complexity
+- Insert: O(n)
+- Search: O(n)
+- Delete: O(n)
+- Print (Inorder): O(n)
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:478]: Keep both of them in `## Complexity Analysis` section

File: docs/data-structures/binary-search-tree.mdx [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+            print()
+
+```
+</TabItem>
+</Tabs>
+
+## Sample I/O
+
+
+### Sample input
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:445]: Put a line break

Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:445]: Keep heading-4

File: docs/data-structures/binary-search-tree.mdx (hunk 1 above)
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:464]: Keep heading-4

File: docs/data-structures/binary-search-tree.mdx [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+---
+
+## Introduction
+
+The binary search tree is a **binary tree** in which for every node, all the elements in its left sub-tree are smaller than that node, and all the elements in its right sub-tree are greater than that node. A binary search tree is useful for searching an element in less number of comparisons. Binary search trees are considered to be efficient data structures because they speed up the process of **insertion** and **deletion**.
+
+## Explanation
+
+<p align="center">
+  <img width="500" height="300" src="https://i.imgur.com/aYWkuCT.png" />
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:13]: Did you make this image?

File: docs/data-structures/binary-search-tree.mdx [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+</p>
+
+As you can see in the above diagram, The `root` node is `30`. The `left sub-tree` of the root node consists of nodes `10`, `15` and `20`. All these nodes have `lesser values` than the root node. The `right sub-tree` of the root node consists of nodes `40`, `50` and `60`. All these nodes have `greater values` than the root node. Each of the sub-trees recursively obeys the binary search tree constraint. For example, in the left sub-tree of the root node, `15` is the root and the element in its left sub-tree `(10)` is smaller than 15, while the node present in the right sub-tree `(20)` is greater than the root node's value.
+
+## Algorithm
+
+This section will discuss the operations performed on a binary search tree and the algorithmic steps used for the same.
+
+### *Search Operation in BST*
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:22]: You don't need to keep bold in headings

File: docs/data-structures/binary-search-tree.mdx [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+* If the key value is greater than the current node:
+    * If the current node has a `right child`, then continue the search process in `right subtree`.
+    * If the key value is found, then display "Node Found" and exit the function.
+    * Else, display `"Node not found"`.
+* If the key value is smaller than the current node:
+    * If the current node has a `left child`, then continue the search process in `left subtree`.
+    * If the key value is found, then display "Node Found" and exit the function.
+    * Else, display "Node not found".
+
+### *Insertion Operation in BST*
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:39]: You don't need to keep bold in headings

File: docs/data-structures/binary-search-tree.mdx [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+
+In a binary search tree, the insertion operation is performed as follows:
+
+**Always perform insertion on the leaf node.**
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:43]: You don't need to keep bold in headings

File: docs/data-structures/binary-search-tree.mdx [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+* Else, `create` a `new node` with the given key value.
+* Start at the `root node` as the `current node`.
+* If the new node's key value is `smaller` than the current node's key value:
+    * If the current node has a `left child`, then move to its left child.
+    * Else, `add` the new node as the current's left child.
+* If the new node's key value is `greater` than the current node's key value:
+    * If the current node has a `right child`, then move to its right child.
+    * Else, add the new node as the current's right child.
+
+### *Deletion Operation in BST*
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:56]: You don't need to keep bold in headings

File: docs/data-structures/binary-search-tree.mdx [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+
+The binary search tree is a **binary tree** in which for every node, all the elements in its left sub-tree are smaller than that node, and all the elements in its right sub-tree are greater than that node. A binary search tree is useful for searching an element in less number of comparisons. Binary search trees are considered to be efficient data structures because they speed up the process of **insertion** and **deletion**.
+
+## Explanation
+
+<p align="center">
+  <img width="500" height="300" src="https://i.imgur.com/aYWkuCT.png" />
+</p>
+
+As you can see in the above diagram, The `root` node is `30`. The `left sub-tree` of the root node consists of nodes `10`, `15` and `20`. All these nodes have `lesser values` than the root node. The `right sub-tree` of the root node consists of nodes `40`, `50` and `60`. All these nodes have `greater values` than the root node. Each of the sub-trees recursively obeys the binary search tree constraint. For example, in the left sub-tree of the root node, `15` is the root and the element in its left sub-tree `(10)` is smaller than 15, while the node present in the right sub-tree `(20)` is greater than the root node's value.
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:16]: No need to keep `left sub-tree in backticks. No need to keep `right sub-tree` in backticks. Keep `lesser values` and `greater values` in bold

File: docs/data-structures/binary-search-tree.mdx [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+  ]
+}>
+
+<TabItem value="cpp">
+
+```cpp
+#include <iostream>
+using namespace std;
+
+struct Node{               //Structure is declared which holds Node data and address of left and right child
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:104]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+    struct Node *lchild;
+    int data;
+    struct Node *rchild;
+}*root=NULL;      // root object is declared globally for struct Node
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:108]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+
+void insert(int element)    //Insertion happens here
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:110]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+{
+    struct Node *p=root;   // using 2 pointers for keeping track of Nodes
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:112]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+    else
+    {
+        while(p!=NULL)
+        {
+            q=p;
+            if(p->data==element)
+            {
+                return;
+            }
+            else if(element < p->data)  //if Inserted data is less than current Node data, move to left child
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:130]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 15]
```diff
@@ -0,0 +1,483 @@
...
+            {
+                p=p->lchild;
+            }
+            else{
+                p=p->rchild;    //if Inserted data is greater than or equal to current Node data, move to right child
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:135]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+        if(element > q->data) 
+        {
+            q->rchild=tmp;
+        }
+        else
+            q->lchild=tmp;
+    }
+}
+
+void Search(Node *p,int key)  //Searching operation takes key as input and search for key in BST
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:147]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 17]
```diff
@@ -0,0 +1,483 @@
...
+            else{
+                p=p->rchild;
+            }
+        }
+    }
+    cout << "Element not found "<<endl;
+
+}
+
+void InOrder(Node *p)   //Inorder traversal is used because inorder of BST gives Sorted output
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:174]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 18]
```diff
@@ -0,0 +1,483 @@
...
+    if(p)
+    {
+        InOrder(p->lchild);
+        cout << p->data << " ";
+        InOrder(p->rchild);
+    }
+
+}
+
+int Height(Node *p)  //To find height of corresponding node
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:185]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 19]
```diff
@@ -0,0 +1,483 @@
...
+if(p==NULL)return 0;
+x=Height(p->lchild);
+y=Height(p->rchild);
+return x>y?x+1:y+1;
+}
+
+/*For deleting any Node on BST, Deleted Node has to be replaced with Appropiate Node without affecting the property of BST
+If we are deleting any node, corresponding node should be replaced by either InOrder predecessor or Inorder Successor based on height condition*/
+
+struct Node * InOrderpredecssor(Node *p)  //If height of left child is greater than right child,it will enter this loop
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:197]: Use comments above line of code, not beside it.

File: docs/data-structures/binary-search-tree.mdx [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+{
+    while(p && p->rchild!=NULL)
+    {
+        p=p->rchild;
+    }
+    return p;
+}
+
+struct Node * InOrderSuccessor(Node *p)  //If height of left child is less than right child,it will enter this loop
```
Comment (HarshCasper) [inline docs/data-structures/binary-search-tree.mdx:206]: Use comments above line of code, not beside it.

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: docs/data-structures/binary-search-tree.mdx (hunk 4 above)
Comment (Saatvik21) [reply docs/data-structures/binary-search-tree.mdx:13]: Yup I made this image.

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**